plutus-core/plutus-core/src/PlutusCore/Builtin/KnownTypeAst.hs

```diff
@@ -143,77 +150,89 @@ data family RepHole x
--- 'ToBinds', so handling built-in types in a special way for each of those would be a hassle,
--- especially given the fact that type-level Haskell is not exactly good at computing things.
--- With the 'ElaborateBuiltin' approach we get 'KnownTypeAst', 'ToHoles' and 'ToBinds' for free.
-type ElaborateBuiltin :: forall k. k -> k
-type family ElaborateBuiltin a where
-    ElaborateBuiltin (f x) = ElaborateBuiltin f `TyAppRep` x
-    ElaborateBuiltin f     = BuiltinHead f
+-- | Specifies that the given type is a built-in one and can be embedded into a 'Type'.
+type HasTypeLevel :: forall a. (GHC.Type -> GHC.Type) -> a -> GHC.Constraint
+type HasTypeLevel uni x = KnownTypeAst Void uni (ElaborateBuiltin uni x)
 
--- | A constraint for \"@a@ is a 'KnownTypeAst' by means of being included in @uni@\".
-type KnownBuiltinTypeAst uni a = KnownTypeAst uni (ElaborateBuiltin a)
+-- | Specifies that the given type is a built-in one and its values can be embedded into a 'Term'.
+type HasTermLevel :: forall a. (GHC.Type -> GHC.Type) -> a -> GHC.Constraint
+type HasTermLevel uni = Includes uni
+
+-- | The product of 'HasTypeLevel' and 'HasTermLevel'.
+type HasBothLevel :: forall a. (GHC.Type -> GHC.Type) -> a -> GHC.Constraint
+type HasBothLevel uni x = (uni `HasTypeLevel` x, uni `HasTermLevel` x)
 
-type KnownTypeAst :: forall a. (GHC.Type -> GHC.Type) -> a -> GHC.Constraint
-class KnownTypeAst uni x where
+-- TODO: make it @forall {a}@ once we have that.
+-- (see https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0099-explicit-specificity.rst)
+-- | Convert a Haskell representation of a possibly 0-ary application of a built-in type to
+-- arbitrary types implementing 'KnownTypeAst'.
+mkTyBuiltin :: forall a (x :: a) uni ann tyname. uni `HasTypeLevel` x => ann -> Type tyname uni ann
+mkTyBuiltin ann = ann <$ typeMapNames absurd (toTypeAst $ Proxy @(ElaborateBuiltin uni x))
+{-# INLINE mkTyBuiltin #-}
+
+-- | A constraint for \"@a@ is a 'KnownTypeAst' by means of being included in @uni@\".
+type KnownBuiltinTypeAst :: forall a. GHC.Type -> (GHC.Type -> GHC.Type) -> a -> GHC.Constraint
+type KnownBuiltinTypeAst tyname uni x = AllBuiltinArgs uni (KnownTypeAst tyname uni) x
+
+-- | This class allows one to convert the type-level Haskell representation of a Plutus type into
+-- the corresponding Plutus type. Associated type families are needed to help elaboration.
+--
+-- Depending on the universe converting a Haskell type to a Plutus team can give different results
+-- (e.g. 'Int' can be a built-in type instead of being encoded via built-in 'Integer'), hence this
+-- class takes a @uni@ argument. Plus, elaboration is universe-specific too.
+--
+-- 'KnownTypeAst' class also takes a @tyname@ argument, so that we can define 'mkTyBuiltin' that is
+-- generic over the type of names (in addition to 'TyName' we also have 'TyDeBruijn', so genericity
+-- does help).
```

Comment:
This elaborates on pretty much all the changes in the PR. I generalized over `tyname` because of [this](https://github.com/input-output-hk/plutus/pull/5390#discussion_r1244582581) discussion -- wasn't hard in the end at all, just a bit of straightforward boilerplate.